### DISCUSSION TITLE
Is it possible to update the recurrence of a repeatable job?

### DISCUSSION THREAD

**Discussion by cursedId0l** [taskforcesh/bullmq#2407] [Q&A]
I've read the docs and understand that under the hood repeatable jobs are just delayed jobs. I also see its possible to update the delay of a delayed job while its in the delayed state. Therefore is it possible to update the recurrence of a repeatable job? It doesn't explicitly mention this in the docs.

**Comment by manast** [comment]
For now this is the best answer I can give (ChatGPT):

To update the recurrence of a repeatable job in BullMQ, you essentially have two main approaches: removing the old repeatable job and creating a new one with the updated recurrence pattern, or directly updating the repeatable job if you're managing your jobs at a more granular level through Redis.

### Approach 1: Remove and Add a New Job

This is the most straightforward approach. First, you need to remove the existing repeatable job and then add a new job with the updated recurrence pattern. Here's how you can do it:

1. **Remove the Existing Repeatable Job**

   You can remove a repeatable job by using the `removeRepeatable` method, which requires specifying the repeat options that uniquely identify the repeatable job you want to remove.

   ```javascript
   const repeatableJobKey = {
     // The key properties that identify the repeatable job, such as:
     cron: '0 */1 * * *', // Example of the old CRON expression
     jobId: 'myRepeatableJob', // If you have assigned a custom jobId
     // Include other identifying properties used when the job was added
   };

   queue.removeRepeatable(repeatableJobKey)
     .then(() => console.log('Repeatable job removed'))
     .catch(err => console.error('Failed to remove repeatable job', err));
   ```

2. **Add a New Job with Updated Recurrence**

   After successfully removing the old job, add a new repeatable job with the desired recurrence pattern.

   ```javascript
   queue.add('myRepeatableJob', { /* job data */ }, {
     repeat: {
       cron: '0 */2 * * *' // New CRON expression for updated recurrence
     }
   })
   .then(job => console.log('Added new repeatable job', job.id))
   .catch(err => console.error('Failed to add new repeatable job', err));
   ```

### Approach 2: Direct Redis Manipulation (Advanced)

For advanced users who are comfortable manipulating Redis directly, you can update the job's recurrence directly in Redis. This approach is more complex and requires a good understanding of how BullMQ stores repeatable jobs in Redis. It's not recommended unless you have specific needs that the BullMQ API cannot accommodate.

### Best Practices

- **Validation:** Always validate the new recurrence pattern to ensure it's correct.
- **Error Handling:** Implement robust error handling especially when removing and adding jobs to handle cases where operations might fail.
- **Job Identification:** Be precise in how you identify jobs for removal to avoid accidentally removing the wrong job.

For most use cases, the first approach (remove and add) is recommended for its simplicity and direct support through the BullMQ API. Direct Redis manipulation should be reserved for exceptional cases due to its complexity and the potential for unintended side effects.
